restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java

```diff
@@ -489,12 +498,238 @@ public ReactiveDocumentService(
+      DocumentDB db,
+      String namespace,
+      String collection,
+      String documentId,
+      List<String> processedPath,
+      Object value,
+      ExecutionContext context) {
+    return getDocument(db, namespace, collection, documentId, processedPath, null, context)
+        .map(
+            array -> {
+              JsonNode data = array.getData();
+              if (!data.isArray()) {
+                throw new ErrorCodeRuntimeException(
+                    ErrorCode.DOCS_API_SEARCH_ARRAY_PATH_INVALID,
+                    "The path provided to push to has no array");
+              }
+              ArrayNode arrayData = (ArrayNode) data;
+              if (value == null) {
+                arrayData.addNull();
+              } else if (value instanceof Map || value instanceof List) {
+                arrayData.add(objectMapper.valueToTree(value));
+              } else if (value instanceof Boolean) {
+                arrayData.add((Boolean) value);
+              } else if (value instanceof String) {
+                arrayData.add((String) value);
+              } else if (value instanceof Integer) {
+                arrayData.add(Long.valueOf((Integer) value));
+              } else if (value instanceof Long) {
+                arrayData.add((Long) value);
+              } else {
+                arrayData.add((Double) value);
+              }
+              return arrayData;
+            });
+  }
+
+  /**
+   * Retrieves the specified path in the document as an array, pops the last (i.e. highest-index)
+   * value, and returns both what the array will look like after the pop and the value that was
+   * popped. Note that this doesn't write the array back to the data store.
+   *
+   * @param db
+   * @param namespace
+   * @param collection
+   * @param documentId
+   * @param processedPath
+   * @param context
+   * @return a Pair containing the JSON representation of the array after popping the value and the
+   *     value itself
+   */
+  public Maybe<Pair<ArrayNode, Object>> getArrayAndValueAfterPop(
+      DocumentDB db,
+      String namespace,
+      String collection,
+      String documentId,
+      List<String> processedPath,
+      ExecutionContext context) {
+    return getDocument(db, namespace, collection, documentId, processedPath, null, context)
+        .map(
+            array -> {
+              JsonNode data = array.getData();
+              if (data == null || !data.isArray()) {
+                throw new ErrorCodeRuntimeException(
+                    ErrorCode.DOCS_API_SEARCH_ARRAY_PATH_INVALID,
+                    "The path provided to pop from has no array");
+              }
+              ArrayNode arrayData = (ArrayNode) data;
+              if (arrayData.size() == 0) {
+                throw new ErrorCodeRuntimeException(ErrorCode.DOCS_API_ARRAY_POP_OUT_OF_BOUNDS);
+              }
+              JsonNode value = arrayData.get(arrayData.size() - 1);
```

Comment:
why don't we call remove the `remove` here, it will return back the removed `JsonNode`?